### ISSUE TITLE
Array iteration is slow

### ISSUE THREAD

**Issue by Cody-G** [JuliaMath/FixedPointNumbers.jl#82]
Here's an example:

```julia
julia> f = x->begin 
              i=0;
              for k in eachindex(x)
                   i+=x[k]
              end
       end
(::#21) (generic function with 1 method)

julia> d = rand(N2f14, 100,100,100,100);

julia> @time f(d);
 11.956324 seconds (700.00 M allocations: 10.431 GB, 10.25% gc time)

julia> using Images
julia> @time f(rawview(d));
  0.128216 seconds (1.73 k allocations: 81.318 KB)
```

I'm getting by for now with using `Images.rawview` when iterating through `N2f14` arrays, but I expected that wouldn't be necessary.

**Comment by timholy** [comment]
I was going to say that it's a type instability, you should declare `i = 0.0` because
```julia
julia> x = N2f14(0.9)
0.90002N2f14

julia> typeof(0+x)
Float32
```
so your loop isn't type-stable. But that turns out to be only part of the problem; the more serious problem is that `promote` isn't inlined, and some stuff gets boxed, so this is just dirt-slow. Interestingly, the following is fast:
```julia
function summer(x)
    i = 0.0f0
    for k in eachindex(x)
        i += convert(Float32, x[k])
    end
    i
end
```

**Comment by Cody-G** [comment]
Ah so the promotion was broken for `Float64` but not `Float32`?

I actually tried to find a boxing problem myself before raising the issue, but I couldn't find any evidence of boxing.  Could you comment on how you found the issue?  I tried the approach below, but I don't notice any clues from `@code_warntype`  Similar issues seem to be coming up often lately so I'd like to get better at tracking them down.

```julia
julia> a = rand(N2f14, 3,3,3);
julia> @code_warntype getindex(a, CartesianIndex(3,3,3))
Variables:
  #self#::Base.#getindex
  A::Array{FixedPointNumbers.Normed{UInt16,14},3}
  I::Tuple{CartesianIndex{3}}

Body:
  begin 
      $(Expr(:meta, :propagate_inbounds)) # line 760:
      SSAValue(1) = (Core.getfield)(I::Tuple{CartesianIndex{3}},1)::CartesianIndex{3}
      return (Base.arrayref)(A::Array{FixedPointNumbers.Normed{UInt16,14},3},(Core.getfield)((Core.getfield)(SSAValue(1),:I)::Tuple{Int64,Int64,Int64},1)::Int64,(Core.getfield)((Core.getfield)(SSAValue(1),:I)::Tuple{Int64,Int64,Int64},2)::Int64,(Core.getfield)((Core.getfield)(SSAValue(1),:I)::Tuple{Int64,Int64,Int64},3)::Int64)::FixedPointNumbers.Normed{UInt16,14}
  end::FixedPointNumbers.Normed{UInt16,14}

julia> @code_warntype +(a[1], 1.0)
Variables:
  #self#::Base.#+
  x::FixedPointNumbers.Normed{UInt16,14}
  y::Float64

Body:
  begin 
      return (Core._apply)(Base.+,$(Expr(:invoke, LambdaInfo for promote(::FixedPointNumbers.Normed{UInt16,14}, ::Float64), :(Base.promote), :(x), :(y))))::Float64
  end::Float64

```

**Comment by Cody-G** [comment]
Ahh I see now that `@code_warntype promote(a[1], 0.0)` reveals the boxing, but I guess I'm not sure what made you suspect that `promote` was the problem.

**Comment by timholy** [comment]
I saw that the call to `promote` wasn't inlined in the LLVM, and indeed that there were GC calls. This may be a codegen bug in Julia. Note that 0.6 handled your test function much more gracefully, so I don't think it's worth reporting.

Even on 0.6, I found the performance was better after reverting #80. Promote calls `convert` (twice), and force-inlining it makes `promote` much bigger. In this case it seems that it's better to inline the `promote` even if `convert` has to be dispatched.
